config/config.go

```diff
@@ -0,0 +1,229 @@
+
+func (ch *MySqlConfigHandler) GetConfig() *Config {
+	ch.RLock()
+	defer ch.RUnlock()
+	return ch.Config
+}
+
+func (ch *MySqlConfigHandler) ReloadConfig(filename string, mysqldAddress string, mysqldUser string, tlsInsecureSkipVerify bool, logger log.Logger) error {
+	var host, port string
+	defer func() {
+		if err != nil {
+			configReloadSuccess.Set(0)
+		} else {
+			configReloadSuccess.Set(1)
+			configReloadSeconds.SetToCurrentTime()
+		}
+	}()
+
+	if cfg, err = ini.LoadSources(
+		opts,
+		[]byte("[client]\npassword = ${MYSQLD_EXPORTER_PASSWORD}\n"),
+		filename,
+	); err != nil {
+		return fmt.Errorf("failed to load %s", filename)
+	}
+
+	if host, port, err = net.SplitHostPort(mysqldAddress); err != nil {
+		return fmt.Errorf("failed to parse address: %s", err)
+	}
+
+	if clientSection := cfg.Section("client"); clientSection != nil {
+		if cfgHost := clientSection.Key("host"); cfgHost.String() == "" {
+			cfgHost.SetValue(host)
+		}
+		if cfgPort := clientSection.Key("port"); cfgPort.String() == "" {
+			cfgPort.SetValue(port)
+		}
+		if cfgUser := clientSection.Key("user"); cfgUser.String() == "" {
+			cfgUser.SetValue(mysqldUser)
+		}
+	}
+
+	cfg.ValueMapper = os.ExpandEnv
+	config := &Config{}
+	m := make(map[string]MySqlConfig)
+	for _, sec := range cfg.Sections() {
+		sectionName := sec.Name()
+
+		if sectionName == "DEFAULT" {
+			continue
+		}
+
+		mysqlcfg := &MySqlConfig{
+			TlsInsecureSkipVerify: tlsInsecureSkipVerify,
+		}
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to load config", "section", sectionName, "err", err)
+			continue
+		}
+
+		err = sec.StrictMapTo(mysqlcfg)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to parse config", "section", sectionName, "err", err)
+			continue
+		}
+		if err := mysqlcfg.validateConfig(); err != nil {
+			level.Error(logger).Log("msg", "failed to validate config", "section", sectionName, "err", err)
+			continue
+		}
+
+		m[sectionName] = *mysqlcfg
+	}
+	config.Sections = m
+	if len(config.Sections) == 0 {
+		return fmt.Errorf("no configuration found")
+	}
+	ch.Lock()
+	ch.Config = config
+	ch.Unlock()
+	return nil
+}
+
+func (m MySqlConfig) validateConfig() error {
+	if m.User == "" {
+		return fmt.Errorf("no user specified in section or parent")
+	}
+	if m.Password == "" {
+		return fmt.Errorf("no password specified in section or parent")
+	}
+
+	return nil
+}
+
+func (m MySqlConfig) FormDSN(target string) (string, error) {
+	var dsn, host, port string
+
+	user := m.User
+	password := m.Password
+	if target == "" {
+		host := m.Host
+		port := m.Port
+		socket := m.Socket
+		if socket != "" {
+			dsn = fmt.Sprintf("%s:%s@unix(%s)/", user, password, socket)
+		} else {
+			dsn = fmt.Sprintf("%s:%s@tcp(%s:%d)/", user, password, host, port)
+		}
+	} else {
+		if host, port, err = net.SplitHostPort(target); err != nil {
+			return dsn, fmt.Errorf("failed to parse target: %s", err)
+		}
+		dsn = fmt.Sprintf("%s:%s@tcp(%s:%s)/", user, password, host, port)
```

Comment:
what if a password contains : ? shouldn't we build this more carefully?

Reply:
This has been a long-standing issue with various Go SQL drivers. They all seem to suffer from this parsing problem.

There's https://pkg.go.dev/github.com/go-sql-driver/mysql?utm_source=godoc#Config which should probably be used here.

Reply:
I am wondering if we couldn't leave this for another issue to rework? This PR is already adding quite a lot.

It will require quite a fair amount of work to pass that along to the Exporter and parse parameters, and my time is quite limited right now.

Reply:
That's fine. Let's work on it next.